docs/architecture/ADR-002-universal-share-encoding.md

```diff
@@ -0,0 +1,161 @@
+# ADR 009: Universal Share Encoding
+<!-- disable markdownlint MD010 because Go code snippet uses tabs -->
+<!-- markdownlint-disable MD010 -->
+
+## Changelog
+
+- 2022/9/22: inital draft of InfoReservedByte
+- 2022/9/24: update draft to Universal Share Encoding
+
+## Context
+
+- **nid** (8 bytes): namespace id
+- **reserved** (1 byte): is the location of the first transaction, ISR, or evidence in this share if there is one and `0` if there isn't one
+- **message length** (varint 1 to 10 bytes): is the length of the entire message in bytes
+
+The current reserved namespace (transaction, ISRs, evidence) share format is:<br>`nid (8 bytes) | reserved (1 byte) | share data`
+
+The current unreserved namespace (message) share format is:
+
+- First share of message:<br>`nid (8 bytes) | message length (varint 1 to 10 bytes) | share data`
+- Contiguous share in message:<br>`nid (8 bytes) | share data`
+
+The current share format poses multiple challenges:
+
+1. Clients must have two share parsing implementations (one for reserved namespace shares and one for unreserved namespace shares).
+1. It is difficult to make changes to the share format in a backwards compatible way because clients can't determine which version of the share format an individual share conforms to.
+1. It is not possible for a client that samples a random share to determine if the share is the first share of that namespace or a contiguous share.
+
+## Proposal
+
+Introduce a universal share encoding that applies to both reserved and unreserved share formats:
+
+- First share of namespace (for reserved namespaces) or message (for unreserved namespaces):<br>`nid (8 bytes) | info (1 byte) | message length (varint 1 to 10 bytes) | data`
+- Contiguous shares in namespace:<br>`nid (8 bytes) | info (1 byte) | data`
```

Comment:
Should rather be contiguous shares in message?

Reply:
On second thought, I see why namespace could be confusing in the scenario where multiple messages for the same namespace are included in a block.

Is this clearer?
```suggestion
- Contiguous shares in namespace (for reserved namespaces) or message (for unreserved namespaces):<br>`nid (8 bytes) | info (1 byte) | data`
```